reasoning_gym/logic/knights_knaves.py

```diff
@@ -0,0 +1,458 @@
+            return (not val1) or val2
+        if statement[0] == "<=>":
+            val1 = KnightsKnavesDataset.test_satisfiability(statement[1], assignments)
+            val2 = KnightsKnavesDataset.test_satisfiability(statement[2], assignments)
+            return (val1 and val2) or ((not val1) and (not val2))
+        raise KeyError(f"Unknown statement: {statement}")
+
+    def __getitem__(self, idx: int) -> dict[str, Any]:
+        """
+        Generate a single knights and knaves problem item.
+
+        Args:
+            idx: Index of the item to generate
+
+        Returns:
+            dict containing at least:
+                - question: str (the puzzle in natural language)
+                - answer: str (the solution in text)
+                - metadata: dict (additional problem details)
+        """
+        rng = Random(self.seed + idx if self.seed is not None else None)
+        return self.__generate_problem(rng)
+
+    def __generate_problem(self, rng: Random) -> dict[str, Any]:
+        """
+        Generate a single knights and knaves problem with a unique solution.
+        """
+
+        # Sample a valid problem using the original KKProblemSampler logic
+        # Use the sampler to generate a valid problem
+        sampler = KKProblemSampler(
+            rand_seed=rng.randint(0, 2**32),
+            n_people=self.config.n_people,
+            depth_constraint=self.config.depth_constraint,
+            width_constraint=self.config.width_constraint,
+        )
+        problems = sampler.sample_valid_problems(1, skip_no_solution=True, skip_multiple_solutions=True)
+        problem = problems[0]
+
+        # Format the problem using the original KKProblemFormatter logic
+
+        # Format the problem
+        formatter = KKProblemFormatter(rand_seed=rng.randint(0, 2**32), problem=problem)
+        formatted = formatter.format_problem()
+
+        # Prepare the return dictionary
+        question = formatted["quiz"]
+        answer = formatted["solution_text"]
+        metadata = {
+            "statements": problem["statements"],
+            "solution": problem["solution"],
+            "names": formatted["names"],
+            "knight_knave_terms": formatted["knight_knave"],
+        }
+
+        return {"question": question, "answer": answer, "metadata": metadata}
+
+    def _normalize_answer(self, answer: str) -> set[tuple[str, str]]:
+        """Convert answer string into normalized set of (name, role) tuples"""
+        # Remove common punctuation and standardize spacing
+        answer = answer.lower().strip().replace(".", "").replace(",", "")
+
+        # Split on 'and' or spaces for different formats
+        parts = [p.strip() for p in answer.replace(" and ", " ").split()]
+
+        # Extract name-role pairs
+        assignments = set()
+        current_name = None
+
+        for part in parts:
+            if part in ["is", "a"]:
+                continue
+            if part in ["knight", "knave"]:
```

Comment:
This probably only works if knight & knave are actually used .. e.g. the following is probably not correct, is it?
```
>>> x[123]['answer']
'Lily is a villain, and Emily is a villain.'
>>> x._normalize_answer(x[123]['answer'])
set()
```